TextSelection/Views/MultitouchTextView.swift

```diff
@@ -0,0 +1,225 @@
+                // For normal selections
+                self.removeGestureRecognizer(recognizer)
+                break
+            case is UITapGestureRecognizer.Type:
+                // Not sure what this is
+                // Probably "cancels" selections
+                self.removeGestureRecognizer(recognizer)
+                break
+            case is UIPanGestureRecognizer.Type:
+                // For scrolling
+                let pan = recognizer as! UIPanGestureRecognizer
+                pan.minimumNumberOfTouches = 2
+                break
+            default:
+                let className = String(Mirror(reflecting: recognizer).subjectType)
+                if (className == "UITapAndAHalfRecognizer") {
+                    // For tap and drag
+                    self.removeGestureRecognizer(recognizer)
+                } else if (className == "_UIPreviewInteractionTouchObservingGestureRecognizer") {
+                    // Not sure what this is
+                }
+                break
+            }
+        })
+    }
+    
+    var baseAttributes: [String: AnyObject] {
+        get {
+            return [NSForegroundColorAttributeName: UIColor.darkTextColor()]
+        }
+    }
+    
+    var preSelectionTimer: NSTimer?
+    
+    func clearPreSelectionWithDelay() {
+        self.preSelectionTimer?.invalidate()
+        self.preSelectionTimer = NSTimer.scheduledTimerWithTimeInterval(2.0, target: self, selector: #selector(clearPreSelection), userInfo: nil, repeats: false)
+    }
+    
+    func clearPreSelection() {
+        self.preSelection = nil
+        self.updateViews(true)
+        self.selectionDidChange()
+    }
+    
+    func didTapButton() {
+        if (self.selection != nil && self.preSelection == nil) {
+            self.clearSelection()
+        } else {
+            self.appendPreSelectionToSelection()
+        }
+    }
+    
+    var shouldShowControlButton: Bool {
+        get {
+            return self.selection != nil || self.preSelection != nil
+        }
+    }
+}
+
+extension MultitouchTextView {
+    override func touchesBegan(touches: Set<UITouch>, withEvent event: UIEvent?) {
+        if let point = touches.first?.locationInView(self.textInputView) {
+            self.startHighlighting(point)
+        }
+    }
+    
+    override func touchesMoved(touches: Set<UITouch>, withEvent event: UIEvent?) {
+        if let point = touches.first?.locationInView(self.textInputView) {
+            self.continueHighlighting(point)
+        }
+    }
+    
+    override func touchesEnded(touches: Set<UITouch>, withEvent event: UIEvent?) {
+        if let point = touches.first?.locationInView(self.textInputView) {
+            self.endHighlighting(point)
+        }
+    }
+}
+
+extension MultitouchTextView {
+    func startHighlighting(point: CGPoint) {
+        let index = self.layoutManager.characterIndexForPoint(point, inTextContainer: self.textContainer, fractionOfDistanceBetweenInsertionPoints: nil)
```

Comment:
The methods glyphIndexForPoint and characterIndexForPoint are subtly different. An example given in the documentation is about how ligatures are drawn as one glyph but are in fact two characters and you can insert between the characters, breaking the ligature.